### PR TITLE
NDRS-557: introduce basic libp2p network component

### PR DESCRIPTION
**Pull request by Fraser999**

### CODE REVIEW

File: execution_engine/src/core/tracking_copy/mod.rs [hunk 1]
```diff
@@ -90,7 +90,7 @@ impl Query {
     fn current_path(&self) -> String {
         let mut path = format!("{:?}", self.base_key);
         for name in &self.visited_names {
-            path.push_str("/");
+            path.push('/');
```
Comment (EdHastingsCasperLabs) [inline execution_engine/src/core/tracking_copy/mod.rs:93]: ?

File: execution_engine/src/storage/trie_store/operations/mod.rs [hunk 2]
```diff
@@ -778,12 +778,7 @@ where
                                 return Some(Err(e));
                             }
                         };
-                        debug_assert!({
-                            match &maybe_next_trie {
-                                Some(Trie::Node { .. }) => true,
-                                _ => false,
-                            }
-                        });
+                        debug_assert!({ matches!(&maybe_next_trie, Some(Trie::Node { .. })) });
```
Comment (henrytill) [inline execution_engine/src/storage/trie_store/operations/mod.rs:781]: Is this a clippy recommend after the changing the toolchain?  I like it either way, but holy seemingly unrelated changes...

**Review by EdHastingsCasperLabs — COMMENTED** [review]
Would you consider updating the toolchain as its own PR, so we can focus on the networking piece in a smaller diff?

File: execution_engine/src/core/tracking_copy/mod.rs (hunk 1 above)
Comment (Fraser999) [reply execution_engine/src/core/tracking_copy/mod.rs:93]: New clippy warning

File: execution_engine/src/storage/trie_store/operations/mod.rs (hunk 2 above)
Comment (Fraser999) [reply execution_engine/src/storage/trie_store/operations/mod.rs:781]: Yeah.  I'll do a separate PR for the upgrade as Ed suggests.

**Comment by Fraser999** [comment]
> Would you consider updating the toolchain as its own PR, so we can focus on the networking piece in a smaller diff?

Sure - this was one of the few times I created a PR with more than one commit to try and allow for separating these things.  But I'll create a PR for just that change now
